### PR TITLE
Update AAA module

### PR DESCRIPTION
**Pull request by stalabi1**
##### SUMMARY

I updated the AAA module authentication implementation to support ordered authentication. This is a breaking change that was reverted in #269. 

Please see [model PR](https://github.com/ansible-network/resource_module_models/pull/209).

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

sonic_aaa

##### OUTPUT

[regression-2024-07-23-14-56-54.html.pdf](https://github.com/user-attachments/files/16354640/regression-2024-07-23-14-56-54.html.pdf)
[diff_output.log](https://github.com/user-attachments/files/15978829/diff_output.log)
[merge_aaa.yaml.txt](https://github.com/user-attachments/files/15978972/merge_aaa.yaml.txt)
[check_merged.log](https://github.com/user-attachments/files/15978980/check_merged.log)
[delete_aaa_yaml.txt](https://github.com/user-attachments/files/15978984/delete_aaa_yaml.txt)
[check_delete.log](https://github.com/user-attachments/files/15979020/check_delete.log)
[facts_gathering.log](https://github.com/user-attachments/files/15979027/facts_gathering.log)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

### CODE REVIEW

**Review by kerry-meyer — APPROVED** [review]
The proposed changes and corresponding test results look good.

Thank you for re-posting these changes to bring the AAA module into compliance with the intended SONiC functionality for AAA.

Approved.

File: plugins/modules/sonic_aaa.py [hunk 1]
```diff
@@ -64,6 +64,7 @@
               default_auth:
                 description:
                   - Specifies order to authenticate aaa login methods
+                version_added: 2.5.0
```
Comment (kerry-meyer) [inline plugins/modules/sonic_aaa.py:67]: Please make this 3.0.0 (because it is part of a "breaking" change).

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
I haven't finished reviewing this PR, but I'm posting my current comments so they can be addressed in parallel with adding support for the new AAA 'name-service' attributes.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 2]
```diff
@@ -220,135 +229,250 @@ def _state_overridden(self, want, have):
...
-                    diff_authentication['data'] = diff_data
-                    diff_cfg['authentication'] = diff_authentication
-
-        return diff_cfg
+            auth_method = authentication.get('auth_method')
+            console_auth_local = authentication.get('console_auth_local')
+            failthrough = authentication.get('failthrough')
+
+            cfg_authentication = have.get('authentication')
+            if cfg_authentication:
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/aaa/aaa.py:417]: You are correct that I overlooked the initial check for absence of any AAA configuration and the default return of 'want'. That covers the first question of my comment.

But the handling for empty configuration **subsets** is still needed: Even if there is configuration for 'authorization', there may not be any for 'authentication' (and vice versa). Both of these cases would be covered by initializing cfg_dict to 'want'.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 3]
```diff
@@ -21,25 +22,47 @@
...
+    console_auth_local = command.get('console_auth_local')
+    failthrough = command.get('failthrough')
+    cfg_auth_method = new_conf.get('auth_method')
+    cfg_console_auth_local = new_conf.get('console_auth_local')
+    cfg_failthrough = new_conf.get('failthrough')
+
+    if auth_method and auth_method == cfg_auth_method:
+        new_conf.pop('auth_method')
+    if console_auth_local and console_auth_local == cfg_console_auth_local:
+        new_conf['console_auth_local'] = False
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/aaa/aaa.py:55]: If a default value is specified for console_auth_local, the current 'derive_authentication_delete' handling for this attribute is correct, but if the 'default' is removed, as suggested in my comment in the 'argspec' file, the handling here should be changed accordingly.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 4]
```diff
@@ -220,135 +207,344 @@ def _state_overridden(self, want, have):
...
+        cfg_authentication = have.get('authentication')
+        cfg_authorization = have.get('authorization')
+        cfg_name_service = have.get('name_service')
+
+        if authentication and authentication != cfg_authentication:
+            config_dict['authentication'] = cfg_authentication
+        if authorization and authorization != cfg_authorization:
+            config_dict['authorization'] = cfg_authorization
+        if name_service and name_service != cfg_name_service:
+            config_dict['name_service'] = cfg_name_service
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/aaa/aaa.py:523]: If the above cfg_*** are None or {}, the replace should not happen.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 5]
```diff
@@ -159,52 +155,43 @@ def _state_merged(self, diff):
...
+        mod_commands = []
         requests = []
-        if diff:
-            requests = self.get_create_aaa_request(diff)
-            if len(requests) > 0:
-                commands = update_states(diff, "replaced")
+        replaced_config = self.get_replaced_config(want, have)
+
+        if replaced_config:
+            is_delete_all = replaced_config == have
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/aaa/aaa.py:181]: There are many lists in args like have, want and replaced_config, if the lists are not sorted, The comparison "==" does nor work properly???

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 6]
```diff
@@ -220,135 +207,344 @@ def _state_overridden(self, want, have):
         requests = []
 
         if have and have != want:
-            del_requests = self.get_delete_all_aaa_request(have)
+            is_delete_all = True
```
Comment (mingjunzhang2019) [inline plugins/module_utils/network/sonic/config/aaa/aaa.py:210]: The comment on line#209: see above comment for _state_replaced.

**Review by mingjunzhang2019 — APPROVED** [review]

File: plugins/module_utils/network/sonic/argspec/aaa/aaa.py [hunk 7]
```diff
@@ -42,16 +42,27 @@ def __init__(self, **kwargs):
...
-                                    'elements': 'str'
-                                },
-                            },
-                            'type': 'dict'
+                        'auth_method': {
+                            'choices': ['ldap', 'local', 'radius', 'tacacs+'],
+                            'elements': 'str',
+                            'type': 'list'
+                        },
+                        'console_auth_local': {'default': False, 'type': 'bool'},
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/argspec/aaa/aaa.py]: Verified.

File: plugins/module_utils/network/sonic/facts/aaa/aaa.py [hunk 8]
```diff
@@ -9,22 +9,24 @@
 for a given resource, parsed, and the facts tree is populated
 based on the configuration.
 """
-
-from __future__ import absolute_import, division, print_function
+from __future__ import (absolute_import, division, print_function)
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/aaa/aaa.py]: Verified

File: plugins/module_utils/network/sonic/facts/aaa/aaa.py [hunk 9]
```diff
@@ -45,61 +47,79 @@ def __init__(self, module, subspec='config', options='options'):
...
-        data = {}
-        if ('openconfig-system:aaa' in response[0][1]):
-            if ('authentication' in response[0][1]['openconfig-system:aaa']):
-                if ('config' in response[0][1]['openconfig-system:aaa']['authentication']):
-                    data = response[0][1]['openconfig-system:aaa']['authentication']['config']
-        return data
-
     def populate_facts(self, connection, ansible_facts, data=None):
-        """ Populate the facts for aaa
+        """ Populate the facts for qos_pfc
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/aaa/aaa.py]: Verified.

File: plugins/module_utils/network/sonic/facts/aaa/aaa.py [hunk 10]
```diff
@@ -45,66 +47,79 @@ def __init__(self, module, subspec='config', options='options'):
...
+            if auth_method:
+                authentication_dict['auth_method'] = auth_method
+            if console_auth_local:
+                authentication_dict['console_auth_local'] = console_auth_local
+            if failthrough:
+                authentication_dict['failthrough'] = bool_dict[failthrough]
+            if authentication_dict:
+                config_dict['authentication'] = authentication_dict
+
+        # Authoriation configuration handling
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/aaa/aaa.py]: Verified

File: plugins/module_utils/network/sonic/facts/aaa/aaa.py [hunk 11]
```diff
@@ -45,66 +47,79 @@ def __init__(self, module, subspec='config', options='options'):
...
+            facts['aaa'] = remove_empties(params['config'])
         ansible_facts['ansible_network_resources'].update(facts)
         return ansible_facts
 
-    def render_config(self, spec, conf):
-        """
-        Render config as dictionary structure and delete keys
-          from spec for null values
+    def get_config(self, module, path, key_name):
+        """Retrive OC configuration from device"""
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/facts/aaa/aaa.py]: Verified.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py [hunk 12]
```diff
@@ -220,135 +229,250 @@ def _state_overridden(self, want, have):
...
+                authentication_dict = {}
+                cfg_auth_method = cfg_authentication.get('auth_method')
+                cfg_console_auth_local = cfg_authentication.get('console_auth_local')
+                cfg_failthrough = cfg_authentication.get('failthrough')
+
+                if auth_method and auth_method != cfg_auth_method:
+                    authentication_dict['auth_method'] = auth_method
+                if console_auth_local is not None and console_auth_local != cfg_console_auth_local:
+                    authentication_dict['console_auth_local'] = console_auth_local
+                if failthrough and failthrough != cfg_failthrough:
```
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/aaa/aaa.py]: Verified.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/aaa/aaa.py]: Verified.

File: plugins/module_utils/network/sonic/config/aaa/aaa.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/aaa/aaa.py:436]: I have checked the current handling for the cases mentioned in this issue. The current handling looks fine and fully addresses the requirements for these cases.

Resolved.

**Review by kerry-meyer — APPROVED** [review]
Now that the 2.x branch sync is done, we can proceed with merging this revised AAA resource module.

I have finished checking the contents of this PR. All proposed changes and corresponding test results look good.

Thank you for providing this revised version of the AAA resource module to better fit the intent and current operation of the AAA feature in SONiC.

Approved.

**Pull request merged**